drivers/storage/portworx/component/autopilot.go

```diff
@@ -265,26 +291,109 @@ func (c *autopilot) createConfigMap(
 	return err
 }
 
+// createServiceAccountForOCP creates service account for OCP Prometheus
+// autopilot to access Prometheus metrics
+// This is required for OCP 4.16 and above
+func (c *autopilot) createServiceAccountForOCP(namespace string, ownerRef *metav1.OwnerReference) error {
+	return k8sutil.CreateOrUpdateServiceAccount(
+		c.k8sClient,
+		&v1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            AutopilotPrometheusServiceAccountName,
+				Namespace:       namespace,
+				OwnerReferences: []metav1.OwnerReference{*ownerRef},
+			},
+		},
+		ownerRef,
+	)
+}
+
+// createClusterRoleBindingForOCP creates cluster role binding for OCP Prometheus
+// autopilot to access Prometheus metrics requires cluster-monitoring-view role
+// This is required for OCP 4.16 and above
+func (c *autopilot) createClusterRoleBindingForOCP(namespace string, ownerRef *metav1.OwnerReference) error {
+	return k8sutil.CreateOrUpdateClusterRoleBinding(
+		c.k8sClient,
+		&rbacv1.ClusterRoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: AutopilotPrometheusClusterRoleBindingName,
+			},
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: "rbac.authorization.k8s.io",
+				Kind:     "ClusterRole",
+				Name:     OpenshiftClusterRoleName,
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind:      "ServiceAccount",
+					Name:      AutopilotPrometheusServiceAccountName,
+					Namespace: namespace,
+				},
+			},
+		},
+	)
+}
+
 func (c *autopilot) createSecret(clusterNamespace string, ownerRef *metav1.OwnerReference) error {
 
-	token, cert, err := c.getPrometheusTokenAndCert()
+	ocp_416, err := pxutil.IsSupportedOCPVersion(c.k8sClient, pxutil.Openshift_4_16_version)
 	if err != nil {
 		return err
 	}
 
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            AutopilotSecretName,
+			Namespace:       clusterNamespace,
+			OwnerReferences: []metav1.OwnerReference{*ownerRef},
+		},
+	}
+
+	if ocp_416 {
+		// OCP 4.16 and above, secret is created by default if autopilot is enabled
+		// if secret is already created, return
+		if c.isCreated {
+			return nil
+		}
+
+		// check if secret already exists, this is first time creation
+		// or secret is created in last reconcilation loop
+		err := k8sutil.GetSecret(c.k8sClient, secret.Name, secret.Namespace, secret)
+		if err != nil {
+			return err
+		}
+
+		// if secret already exists, return
+		// once secret is created, it should not be updated
+		// if secret is deleted, it will be created if operator is restarted
+		if secret.Data != nil {
+			c.isSecretCreated = true
+			return nil
+		}
+
+		// create secret with service account token
+		secret.Type = v1.SecretTypeServiceAccountToken
```

Comment:
You can't change the secret type of the existing secret. I tried that and got this error
```
# * type: Invalid value: "kubernetes.io/service-account-token": field is immutable
```
I think it will be cleaner to just delete the existing secret and create a brand new one.

Reply:
IIUC, this will create a permanent token that is valid forever. Ideally, we should create a limited duration token and keep updating it before the token expires. In PR https://github.com/libopenstorage/operator/pull/1582 , we are using a CreateToken API to create an expiring token.

Reply:
updated